protocol.html

```diff
@@ -559,6 +572,40 @@ <h3 property="schema:name">Resource Containment</h3>
                   <p id="server-hierarchical-containment">There is a 1-1 correspondence between containment triples and relative reference within the path name hierarchy. [<a href="https://github.com/solid/specification/issues/98#issuecomment-547506617" rel="cito:citesAsSourceDocument">Source</a>]. It follows that all resources are discoverable from a container and that it is not possible to create orphan resources. [<a href="https://github.com/solid/specification/issues/97#issuecomment-547459396" rel="cito:citesAsSourceDocument">Source</a>]</p>
 
                   <p><span about="" id="server-basic-container" rel="spec:requirement" resource="#server-basic-container"><span property="spec:statement">The representation and behaviour of containers in Solid corresponds to LDP Basic Container and <span rel="spec:requirementLevel" resource="spec:MUST">MUST</span> be supported by <span rel="spec:requirementSubject" resource="spec:Server">server</span>.</span></span> [<a href="https://github.com/solid/specification/issues/47#issuecomment-561675764" rel="cito:citesAsSourceDocument">Source</a>]</p>
+
+                  <section id="authoritative-resource-data" inlist="" rel="schema:hasPart" resource="#authoritative-resource-data">
+                    <h4 property="schema:name">Authoritative Resource Data</h4>
+                    <div datatype="rdf:HTML" property="schema:description">
+                      <p>Container descriptions are not limited to containment triples. To further support client navigation and application interaction, servers can include <a href="#authoritative-information">authoritative information</a> about contained resources as part of the container description.</p>
+
+                      <p><span about="" id="authoritative-contained-resource-data" rel="spec:requirement" resource="authoritative-contained-resource-data"><span property="spec:statement"><span rel="spec:requirementSubject" resource="spec:Server">Servers</span> <span rel="spec:requirementLevel" resource="spec:MUST">MUST</span> include the following information about contained resources as part of the container description, unless that information does not exist or is expensive to determine:</span></span></p>
```

Comment:
Agreed, but how are we going to test this? The "unless" basically means a flag for the test suite then?

Reply:
The condition is an exit strategy for servers that essentially cannot or do not wish to provide the authoritative information. Inapplicable.

The test suite would require a flag if the reports wish to distinguish the outcome of the test for subjects that implement the required feature from those subjects that it is "inapplicable". (EARL takes this into account.)

I've generalised the condition using "inapplicable" and then expanded on it in #authoritative-resource-data-logically-part-of-container-description .

When the group reaches consensus on a more precise requirement (with an exception or otherwise), we can update the specification.

Reply:
I've updated the text in the meantime that hopefully better communicates this:
https://github.com/solid/specification/pull/352/commits/c0d8d4cea9b66a2914814714b134e9b1cc6042fa

Reply:
I have a general problem with this type of formulation, "When the server... it MUST" or "MUST ... unless". I find that it mostly serves to obscure the actual requirement level and complicate the text. It makes testing more complicated, and it does not help interoperability. Either, it is something you can rely on, or you can't.

I think it would be much clearer to use SHOULD in this case, and explain in a non-normative note in which cases it is OK to consider not doing it.

Reply:
I don't think it complicates testing but I'll take your word for it.

Inapplicable is a valid test outcome. It is not always straight forward to have a requirement with a specific requirement level without additional context. "MUST, unless" or "SHOULD" still generally end up in the same category I think.

(MUST unless is not unheard of in specs. LDP has several "MUST (NOT), unless", if that means anything - probably not)

The current text reflects the discussion in issue 343.

Reply:
Yeah, I know it is being formulated that way. It may be OK in a contexts such as HTTP, where specific applications like Solid needs to establish additional constraints anyway, but I feel it gets very awkward with Solid and LDP.

I think that with a MUST, a test failure is clearly an indication of non-compliance, so you need to code into the tests further intentions of the spec as special cases. In the case of a SHOULD, it is fair to assume that a test failure is probably a case of inapplicable. There's a nuance there, even though the end result is indeed the same, the test is inapplicable.

Reply:
I don't think a failure on a SHOULD is necessarily "inapplicable". That should be a "failed". "inapplicable" or "cantTell" (again from EARL) are possible outcome values provided that there is some context thrown in. And then there is the "info" as part of the test result which would probably explain what/why..

Any way, we are discussing two things here 1) readability/correctness of the requirement 2) validity/usefulness of the test reports. I generally agree with you on both accounts. I'm just saying that for "inapplicable" to happen, there needs to be some context given and that's not due to the requirement level - putting aside the fact why MUST or a SHOULD would be preferable for that particular requirement.

Reply:
It shouldn't be a "cantTell", that's for when the outcome of the test is not certain, and you can't fail a SHOULD or MAY test, because it might just be that you decided to not do it, which is OK. We may add something to the harness where implementers may say that they intend to support a SHOULD or MAY, and therefore fail would be OK. But for now, among the EARL options, it leaves inapplicable. Arguably, a "optional" outcome value or something like that might be desireable.

Nevertheless, I acknowledge I might be conflating things, I guess I should hold my fire on those things until we can have a video chat about it, I feel such issues should not hold up 0.9, but I also don't want to delay the test suite over it.

Reply:
I'd like to update the requirement at some point specifying the particular circumstances in which the server may not honour the requirement. Right now it is a bit vague (and not exciting). Some of the mentions currently in the Note may need to be more precise and move into normative language.

Reply:
Clarified the Note in https://github.com/solid/specification/pull/352/commits/f599dd5d9df818ecaa1217359c45ba66a88d5e04 and https://github.com/solid/specification/pull/352/commits/98d2db4ec1d0438cd455163748df413f967cdb36

Deferring to https://github.com/solid/specification/issues/343 for changes to requirement level and applicability.